### PR TITLE
[RW-963][risk=no] Refactor of modifiers code

### PR DESCRIPTION
**Pull request by freemabd1**
This PR allows Classes that implement AbstractQueryBuilder to use buildModifierSql() method if modifiers apply to the criteria type.

### CODE REVIEW

File: api/src/main/java/org/pmiops/workbench/cohortbuilder/querybuilder/AbstractQueryBuilder.java [hunk 1]
```diff
@@ -1,27 +1,186 @@
...
+      Modifier eventDate = getModifier(modifiers, ModifierType.EVENT_DATE);
+      Modifier occurrences = getModifier(modifiers, ModifierType.NUM_OF_OCCURRENCES);
+      modifierSql = buildAgeAtEventAndEventDateModifierSql(queryParams, Arrays.asList(ageAtEvent, eventDate));
+      //Number of Occurrences has to be last because of the group by
+      modifierSql = modifierSql + buildNumOfOccurrencesModifierSql(queryParams, occurrences);
+    }
+    finalSql = MODIFIER_SQL_TEMPLATE.replace("${innerSql}", baseSql) +
+      modifierSql;
+
+    return QueryJobConfiguration
```
Comment (osgsusan) [inline api/src/main/java/org/pmiops/workbench/cohortbuilder/querybuilder/AbstractQueryBuilder.java]: Is it possible to just return the modifiersql string, and put together this finalsql builder in the specific criteria, for consistency with other non-modifier criteria?

File: api/src/main/java/org/pmiops/workbench/cohortbuilder/querybuilder/AbstractQueryBuilder.java [hunk 2]
```diff
@@ -1,27 +1,186 @@
...
+          try {
+            return new SimpleDateFormat("yyyy-MM-dd").parse(date);
+          } catch (ParseException pe) {
+            throw new BadRequestException(String.format(
+              "Please provide valid date for %s.",
+              exceptionText.get(modifier.getName())));
+          }
+        }).collect(Collectors.toList());
+    }
+  }
```
Comment (osgsusan) [inline api/src/main/java/org/pmiops/workbench/cohortbuilder/querybuilder/AbstractQueryBuilder.java]: great work.

File: api/db-cdr/csv/visit_criteria.csv [hunk 3]
```diff
@@ -1,6 +1,6 @@
 id,parent_id,type,subtype,code,name,is_group,is_selectable,est_count,concept_id,domain_id,has_attribute,predefined_attributes
 316400,0,VISIT,,,Emergency Room Visit,1,1,,9203,Visit,0,
-316401,0,VISIT,,,Emergency Room and Inpatient Visit,1,1,,262,Visit,0,
+316401,0,VISIT,,,Emergency Room and Inpatient Visit,0,1,,262,Visit,0,
```
[the remark this answers is not in the document]
Comment (freemabd1) [reply api/db-cdr/csv/visit_criteria.csv]: thanks

File: api/src/main/java/org/pmiops/workbench/cohortbuilder/querybuilder/AbstractQueryBuilder.java (hunk 1 above)
Comment (freemabd1) [reply api/src/main/java/org/pmiops/workbench/cohortbuilder/querybuilder/AbstractQueryBuilder.java]: Yes, I will change. Done.

File: api/src/main/java/org/pmiops/workbench/cohortbuilder/querybuilder/AbstractQueryBuilder.java [hunk 4]
```diff
@@ -1,27 +1,186 @@
...
+        for (String operand : modifier.getOperands()) {
+          String modifierParameter = isAgeAtEvent ? AGE_AT_EVENT_PREFIX : EVENT_DATE_PREFIX +
+            getUniqueNamedParameterPostfix();
+          modifierParamList.add("@" + modifierParameter);
+          queryParams.put(modifierParameter, isAgeAtEvent ?
+            QueryParameterValue.int64(new Long(operand)) : QueryParameterValue.date(operand));
+        }
+        if (isAgeAtEvent) {
+          if (modifierSql.isEmpty()) {
+            modifierSql = AGE_AT_EVENT_JOIN_TEMPLATE + WHERE + AGE_AT_EVENT_SQL_TEMPLATE;
```
[the remark this answers is not in the document]
Comment (freemabd1) [reply api/src/main/java/org/pmiops/workbench/cohortbuilder/querybuilder/AbstractQueryBuilder.java]: Yes, to calculate age at event we need the birthdate month, day and year which resides on the person table.

File: api/src/main/java/org/pmiops/workbench/cohortbuilder/querybuilder/AbstractQueryBuilder.java (hunk 2 above)
Comment (freemabd1) [reply api/src/main/java/org/pmiops/workbench/cohortbuilder/querybuilder/AbstractQueryBuilder.java]: thanks

File: api/src/main/java/org/pmiops/workbench/cohortbuilder/querybuilder/VisitsQueryBuilder.java [hunk 5]
```diff
@@ -70,9 +72,11 @@ public QueryJobConfiguration buildQueryJobConfig(QueryParameters inputParameters
     }
 
     // Combine the parent and child queries, or just use the one
-    String finalSql = queryParts.size() > 1 ?
+    String visitSql = queryParts.size() > 1 ?
       String.join(UNION_TEMPLATE, queryParts) : queryParts.get(0);
 
+    String finalSql = buildModifierSql(visitSql, queryParams, inputParameters.getModifiers());
```
Comment (osgsusan) [inline api/src/main/java/org/pmiops/workbench/cohortbuilder/querybuilder/VisitsQueryBuilder.java]: Excellent.

**Pull request merged**